### PR TITLE
Updated docs for new WSGI Server options 

### LINKED ISSUE #4802: Update the WSGI entrypoint to not use paste.deploy
TODO

paste.deploy loadapp

#### Story points

### PR DESCRIPTION
**Pull request by kowh-ai**
A fairly extensive re-write (of the deployment doc) to make sure the emphasis is on using uwsgi as the preferred (and supported) WSGI Server (ie: not Apache) for CKAN 2.9. While uwsgi is the preferred WSGI Server, both gunicorn and Apache could be used, and also with the same wsgi script file. NGINX is to be used as the proxy server. The nginx configuration is also included in the revised deployment doc. All 3 WSGI Server options are documented on the wiki: 

https://github.com/ckan/ckan/wiki/CKAN-2.9---New-Web-Server-options

The WSGI Server configuration is very vanilla and does not include any SSL configuration. One of the requirements was the Python virtual environment was to be configured at the WSGI server level rather than within the wsgi script. I have purposely taken out the troubleshooting section as this was very Apache specific. 

Fixes #4802 

### Proposed fixes:
Updated deployment doc and config files included...

### Features:

- [ ] includes tests covering changes
- [ X] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

**Comment by kowh-ai** [comment]
Wiki Page for CKAN 2.9 web server options: https://github.com/ckan/ckan/wiki/CKAN-2.9---New-Web-Server-options

**Comment by kowh-ai** [comment]
@amercader @Zharktas  - Iv'e updated the documentation and added the files now - The build errors refer to wsgi.py which works fine - This can now be tested - thanks for your help guys!

**Comment by kowh-ai** [comment]
@amercader @Zharktas - Iv'e tested a file upload - works OK

**Comment by smotornyuk** [comment]
@clauded , due to module function changes, WSGI script for v2.8 need to be changed a bit (note, i'm using `config.ini` instead of `ckan.ini`:
```
import os
from ckan.config.middleware import make_app

from ckan.lib.cli import _get_config

from logging.config import fileConfig as loggingFileConfig

config_filepath = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'config.ini')
abspath = os.path.join(os.path.dirname(os.path.abspath(__file__)))
loggingFileConfig(config_filepath)
config = _get_config(config_filepath)
application = make_app(config.global_conf, **config.local_conf)
```
and slightly different ini file(i'm not using `callable` and `module`):
```
[uwsgi]
http            =  127.0.0.1:8080
uid             =  www-data
guid            =  www-data
wsgi-file       =  /home/sergey/Projects/gsq/wsgi.py
virtualenv      =  /home/sergey/Projects/gsq
master          =  true
pidfile         =  /tmp/%n.pid
harakiri        =  50
max-requests    =  5000
vacuum          =  true
```

And yes, i'm using different folders, but it shouldn't be an issue. With this, `uwsgi -i wsgi.ini` spins up CKAN instance at http://127.0.0.1:8080

**Comment by clauded** [comment]
@smotornyuk , thanks. I had to make a couple of adjustements to ckan-uwsgi.ini since we are using a ckan.root_path (/recherche):

```
mount              = /recherche=/etc/ckan/default/wsgi.py
manage-script-name = true

```
Also, something similar to this could be useful:

```
req-logger         = file:/var/log/uwsgi/ckan_access.log
logger             = file:/var/log/uwsgi/ckan_access.log
```

Finally, I had to change **config.ini** to **production.ini** in wsgi.py.

**Comment by smotornyuk** [comment]
I've tested and reviewed it  - it can be merged right after fixes for failing tests. I think following patch will solve all the problems:
```diff
diff --git a/wsgi.py b/wsgi.py
index 0392467da..36f289d26 100644
--- a/wsgi.py
+++ b/wsgi.py
@@ -1,9 +1,12 @@
+# -*- coding: utf-8 -*-
+
 import os
 from ckan.config.middleware import make_app
 from ckan.cli import CKANConfigLoader
 from logging.config import fileConfig as loggingFileConfig
-config_filepath = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'ckan.ini')
+config_filepath = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)), u'ckan.ini')
 abspath = os.path.join(os.path.dirname(os.path.abspath(__file__)))
 loggingFileConfig(config_filepath)
 config = CKANConfigLoader(config_filepath).get_config()
-application = make_app(config)
\ No newline at end of file
+application = make_app(config)
```

**Comment by kowh-ai** [comment]
Thanks @smotornyuk - just the Python2 test failing now: **"AssertionError: Unprefixed string literals"** - shall I fix this?

**Comment by smotornyuk** [comment]
Nope, i found out, that i can edit files using GitHub UI:)

File: doc/maintaining/installing/deployment.rst [hunk 1]
```diff
@@ -4,169 +4,166 @@ Deploying a source install
...
-        # Pass authorization info on (needed for rest api).
-        WSGIPassAuthorization On
+-----------------------------------
+4. Install Supervisor for the uwsgi
+-----------------------------------
 
-        # Deploy as a daemon (avoids conflicts between CKAN instances).
-        WSGIDaemonProcess ckan_default display-name=ckan_default processes=2 threads=15
+Install Supervisor_ (a Process Control System) used to control starting, stopping the 
+uwsgi or gunicorn servers::
```
Comment (wardi) [inline doc/maintaining/installing/deployment.rst:103]: uwsgi server

File: doc/maintaining/installing/deployment.rst [hunk 2]
```diff
@@ -4,169 +4,166 @@ Deploying a source install
...
 
 
 ----------------
-2. Install Nginx
+1. Install Nginx
 ----------------
 
-Install Nginx_ (a web server) which will proxy the content from Apache_ and add
-a layer of caching::
+Install NGINX_ (a web server) which will proxy the content from one of the WSGI Servers
```
Comment (wardi) [inline doc/maintaining/installing/deployment.rst:27]: uwsgi

**Comment by clauded** [comment]
Just wandering if it would be better to use a systemd service instead of supervisor to control CKAN?

**Comment by kowh-ai** [comment]
@clauded - yeah pro's and con's for both I think. Supervisor is pretty reliable, permissions could be configured for non-root users to drop a conf file into the conf.d/ directory, may be easier to adapt the CKAN docs for non-Linux systems, can also be used for CKAN background jobs - in fact thats what the CKAN Docker install does...but yes it is 1 more piece of kit to break

**Pull request merged**